pkg/virt-launcher/virtwrap/converter/vcpu/vcpu.go

```diff
@@ -0,0 +1,282 @@
+	threads := map[int]struct{}{}
+	visited := map[uint32]struct{}{}
+	cores = [][][]uint32{}
+	for _, cpu := range cpuSet {
+		threads[cpu] = struct{}{}
+	}
+	for _, cell := range nodeTopology.NumaCells {
+		var coresOnCell [][]uint32
+		for _, cpu := range cell.Cpus {
+			if _, exists := visited[cpu.Id]; exists {
+				continue
+			}
+			core := []uint32{}
+			if len(cpu.Siblings) == 0 {
+				visited[cpu.Id] = struct{}{}
+				if _, exists := threads[int(cpu.Id)]; exists {
+					core = append(core, cpu.Id)
+				}
+			} else {
+				for _, thread := range cpu.Siblings {
+					visited[thread] = struct{}{}
+					if _, exists := threads[int(thread)]; exists {
+						core = append(core, thread)
+					}
+				}
+			}
+			coresOnCell = append(coresOnCell, core)
+		}
+		cores = append(cores, coresOnCell)
+	}
+	return cores
+}
+
+func (p *cpuPool) FitCores() (cpuTune *api.CPUTune, err error) {
+	threads, remaining := p.fitCores(p.cores)
+
+	if remaining > 0 {
+		if p.allowCellCrossing {
+			return nil, fmt.Errorf("not enough exclusive threads provided, could not fit %v core(s)", remaining)
+		} else {
+			return nil, fmt.Errorf("could not fit %v core(s) without crossing numa cell boundaries for individual cores", remaining)
+		}
+	}
+	cpuTune = &api.CPUTune{}
+	for idx, hostThread := range threads {
+		vcpupin := api.CPUTuneVCPUPin{}
+		vcpupin.VCPU = uint32(idx)
+		vcpupin.CPUSet = strconv.Itoa(int(hostThread))
+		cpuTune.VCPUPin = append(cpuTune.VCPUPin, vcpupin)
+	}
+	return cpuTune, nil
+}
+
+func (p *cpuPool) fitCores(coreCount int) (threads []uint32, remainingCores int) {
+	remainingCores = coreCount
+	assignedThreads, remainingCores := p.fitCPUBound(remainingCores)
+	threads = append(threads, assignedThreads...)
+	if remainingCores == 0 {
+		return threads, 0
+	}
+	assignedThreads, remainingCores = p.fitCellBound(remainingCores)
+	threads = append(threads, assignedThreads...)
+	if remainingCores == 0 {
+		return threads, 0
+	}
+	if p.allowCellCrossing {
+		assignedThreads, remainingCores = p.fitUnbound(remainingCores)
+		threads = append(threads, assignedThreads...)
+		if remainingCores == 0 {
+			return threads, 0
+		}
+	}
+	return threads, remainingCores
+}
+
+func (p *cpuPool) FitThread() (thread uint32, err error) {
+	t := p.fitThread()
+	if t == nil {
+		return 0, fmt.Errorf("no remaining unassigned threads")
+	}
+	return *t, nil
+}
+
+func (p *cpuPool) fitCPUBound(requested int) (threads []uint32, remainingCores int) {
+	for _, cell := range p.cells {
+		for {
+			if cell.IsEmpty() {
+				break
+			}
+			junk := cell.GetCPUBoundJunk()
+			if len(junk) > 0 {
+				threads = append(threads, junk...)
+				requested--
+			} else {
+				// go to the next cell
+				break
+			}
+			if requested == 0 {
+				return threads, 0
+			}
+		}
+	}
+	return threads, requested
+}
+
+func (p *cpuPool) fitCellBound(requested int) (threads []uint32, remainingCores int) {
+	for _, cell := range p.cells {
+		for {
+			if cell.IsEmpty() {
```

Comment:
This looks like duplicated logic from above.
Could move into a function taking a `type junkAllocator func() []uint32`.
Like 
```
for _, cell := range p.cells {
  threads, requested = getCellJunk(requested, threads, cell.GetCellBoundJunk)
  if requested < 1 {
    return threads, 0
  }
}
```
Correct me if I'm wrong :-)

Reply:
Done. I chose a slightly different way to avoid having to pass in and out expanding fields. In general if two small code parts are almost identical I sometimes prefer keeping it separate to reduce the magic, but de-duplicated now.